sparse/local.go

```diff
@@ -0,0 +1,167 @@
+package sparse
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// SyncLocalFile syncs a local file from sourceFilePath to targetFilePath.
+// It ensures the file size is a multiple of the Block constant before syncing.
+func SyncLocalFile(sourceFilePath, targetFilePath string) error {
+	log := logrus.WithFields(logrus.Fields{
+		"sourceFilePath": sourceFilePath,
+		"targetFilePath": targetFilePath,
+	})
+
+	fileInfo, err := os.Stat(sourceFilePath)
+	if err != nil {
+		log.WithError(err).Error("Failed to get file info of source file")
+		return err
+	}
+
+	fileSize := fileInfo.Size()
+	if fileSize%Blocks != 0 {
+		return fmt.Errorf("invalid file size %v for local file sync", fileSize)
+	}
+
+	log = log.WithField("fileSize", fileSize)
+	log.Info("Syncing file")
+
+	return syncLocalContent(sourceFilePath, targetFilePath)
+}
+
+func syncLocalContent(sourceFilePath, targetFilePath string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "failed to sync content for source file %v", sourceFilePath)
+	}()
+
+	syncBatchSize := defaultSyncBatchSize
+
+	local := newSyncLocal(sourceFilePath, targetFilePath, syncBatchSize)
+	defer local.close()
+
+	syncStartTime := time.Now()
+
+	err = local.syncContent()
+	if err != nil {
+		return err
+	}
+
+	log := logrus.WithFields(logrus.Fields{
+		"sourceFilePath": sourceFilePath,
+		"targetFilePath": targetFilePath,
+		"elapsed":        time.Since(syncStartTime).Seconds(),
+	})
+	log.Info("Finished syncing file")
+
+	return nil
+}
+
+type syncLocal struct {
+	sourceFilePath string
+	targetFilePath string
+
+	sourceFileIo *DirectFileIoProcessor
+	targetFileIo *DirectFileIoProcessor
+
+	syncBatchSize  int64
+}
+
+func newSyncLocal(sourceFilePath, targetFilePath string, syncBatchSize int64) *syncLocal {
+	return &syncLocal{
+		sourceFilePath: sourceFilePath,
+		targetFilePath: targetFilePath,
+		syncBatchSize:  syncBatchSize,
+	}
+}
+
+func (local *syncLocal) close() {
+	local.sourceFileIo.Close()
+	local.targetFileIo.Close()
+}
+
+func (local *syncLocal) syncContent() error {
+	sourceFileIo, err := NewDirectFileIoProcessor(local.sourceFilePath, os.O_RDONLY, 0)
+	if err != nil {
+		return err
+	}
+	defer sourceFileIo.Close()
+
+	local.sourceFileIo = sourceFileIo
+
+	targetFileIo, err := NewDirectFileIoProcessor(local.targetFilePath, os.O_RDWR|os.O_CREATE, 0)
+	if err != nil {
+		return err
+	}
+	local.targetFileIo = targetFileIo
```

Comment:
Can we open the two files inside `newSyncLocal()`? Then, they can be closed in `local.close()`.